sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -871,6 +871,72 @@ class Column(val expr: Expression) extends Logging {
    */
   def getItem(key: Any): Column = withExpr { UnresolvedExtractValue(expr, Literal(key)) }
 
+  // scalastyle:off line.size.limit
+  /**
+   * An expression that adds/replaces field in `StructType` by name.
+   *
+   * {{{
+   *   val df = sql("SELECT named_struct('a', 1, 'b', 2) struct_col")
+   *   df.select($"struct_col".withField("c", lit(3)))
```

Comment:
Have any of you try to run these examples? The optimizer ConstantFolding rule will break these examples.

Reply:
cc @fqaiser94 @dbtsai @cloud-fan

Reply:
weird, we have tests to cover these examples. @fqaiser94 can you take a look?

Reply:
I failed to write a test case to cover this scenario, my bad. 
And yea, I just tried this example again, and I can see that it fails. 
The issue is that I `override foldable` for this `Unevaluable` Expression. And so, when `foldable` returns true, Spark tries to evaluate the expression and it fails at that point. 
I kind-of realized this as well recently and in my PR for `dropFields` [here](https://github.com/apache/spark/pull/29322/files#diff-c1758d627a06084e577be0d33d47f44eL566), I've fixed the issue (basically i just don't `override foldable` anymore). 
I guess I should submit a follow-up PR to fix this immediately with associated unit tests?

Reply:
Yes, please

Reply:
Thanks for raising the issue @gatorsmile 
I've created a [JIRA](https://issues.apache.org/jira/browse/SPARK-32521) and [PR](https://github.com/apache/spark/pull/29338) to address the issue.